### PR TITLE
[RFC FS-1113] Additions to collection functions

### PR DESCRIPTION
**Pull request by uxsoft**
Suggestion approved in principle: [https://github.com/fsharp/fslang-suggestions/issues/1047](https://github.com/fsharp/fslang-suggestions/issues/1047)

RFC FS-1113 https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1113-insert-remove-update-functions.md

- [x] Map.Keys, Map.Values, Map.keys implementation
- [x] Array removeAt, removeManyAt, updateAt, insertAt, insertManyAt implementation
- [x] List removeAt, removeManyAt, updateAt, insertAt, insertManyAt implementation
- [x] Seq removeAt, removeManyAt, updateAt, insertAt, insertManyAt implementation
- [x] Map tests
- [x] Array tests
- [x] List tests
- [x] Seq tests
- [x] Map benchmark
- [x] Array benchmark
- [x] List benchmark
- [x] Seq benchmark
- [x] Map xml docs
- [x] Array xml docs
- [x] List xml docs
- [x] Seq xml docs

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=11888) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=11888) before we can accept your contribution.<br/><br/>:x: uxsoft [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=11888)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=11888) it.</sub>

File: src/fsharp/FSharp.Core/array.fs [hunk 1]
```diff
@@ -1270,6 +1270,96 @@ namespace Microsoft.FSharp.Collections
...
 
+        [<CompiledName("RemoveAt")>]
+        let removeAt (index: int) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index >= source.Length then invalidArg "index" "index must be within bounds of the array"
+            
+            let length = source.Length - 1
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+
+            for i in [0..length] do
```
Comment (AviAvni) [inline src/fsharp/FSharp.Core/array.fs:1281]: Remoce the braket no need to create lisr same for all methods

Comment (AviAvni) [inline src/fsharp/FSharp.Core/array.fs:1281]: You can start from index

File: src/fsharp/FSharp.Core/array.fs [hunk 2]
```diff
@@ -1270,6 +1270,96 @@ namespace Microsoft.FSharp.Collections
...
+        
+        [<CompiledName("InsertAt")>]
+        let insertAt (index: int) (value: 'T) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index > source.Length then invalidArg "index" "index must be within bounds of the array"
+            
+            let length = source.Length + 1
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+
+            for i in [0..length] do
```
Comment (AviAvni) [inline src/fsharp/FSharp.Core/array.fs:1340]: You delete the i value before copy to i+1

File: src/fsharp/FSharp.Core/array.fs (hunk 1 above)
Comment (dsyme) [reply src/fsharp/FSharp.Core/array.fs:1281]: Yes will be much faster without

File: src/fsharp/FSharp.Core/array.fs [hunk 3]
```diff
@@ -1270,6 +1270,96 @@ namespace Microsoft.FSharp.Collections
                 let count' = Operators.min count len
                 Microsoft.FSharp.Primitives.Basics.Array.subUnchecked 0 count' array
 
+        [<CompiledName("RemoveAt")>]
+        let removeAt (index: int) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index >= source.Length then invalidArg "index" "index must be within bounds of the array"
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs:1276]: The current spec in https://github.com/dotnet/fsharp/pull/11772#issuecomment-886225737 says "if it's outside the range it is ignored".  However I'm open to discussion on this

File: src/fsharp/FSharp.Core/map.fs [hunk 4]
```diff
@@ -933,3 +937,11 @@ module Map =
     [<CompiledName("Count")>]
     let count (table: Map<_, _>) =
         table.Count
+
+    [<CompiledName("Keys")>]
+    let keys (table: Map<_, _>) = 
+        seq { for kvp in table -> kvp.Key }
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/map.fs:943]: Use `table.Keys`

File: src/fsharp/FSharp.Core/map.fs [hunk 5]
```diff
@@ -933,3 +937,11 @@ module Map =
...
+
+    [<CompiledName("Values")>]
+    let values (table: Map<_, _>) = 
+        seq { for kvp in table -> kvp.Value }
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/map.fs:947]: Use table.Values.  Note this will include duplicate values

File: src/fsharp/FSharp.Core/map.fsi [hunk 6]
```diff
@@ -85,6 +85,9 @@ namespace Microsoft.FSharp.Collections
         /// <returns><c>true</c> if the value is present, <c>false</c> if not.</returns>
         member TryGetValue: key:'Key * [<System.Runtime.InteropServices.Out>] value:byref<'Value> -> bool
 
+        member Keys : 'Key seq
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/map.fsi:88]: Documentation for these please

Comment (dsyme) [reply src/fsharp/FSharp.Core/map.fsi:88]: Use `seq<'Key>` in FSharp.Core please

File: src/fsharp/FSharp.Core/seq.fs [hunk 7]
```diff
@@ -1488,3 +1488,72 @@ namespace Microsoft.FSharp.Collections
                                 [|SR.GetString SR.inputMustBePositive; count|]
             mkDelayedSeq (fun () ->
                 source |> toArray |> Array.splitInto count :> seq<_>)
+
+        [<CompiledName("RemoveAt")>]
+        let removeAt (index: int) (source: 'T seq) : 'T seq =
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/seq.fs:1493]: Use `seq<'T>` in FSHarp.Core please

Comment (dsyme) [inline src/fsharp/FSharp.Core/seq.fs:1493]: We need to consider the spec w.r.t. out-of-range indexes, and whether it differs from Array and List.  Here for overly-large indexes the out-of-range can't be discovered until the end of iteration of the list, and it would then seem wrong to raise an exception at the very end of iteration

File: src/fsharp/FSharp.Core/list.fs [hunk 8]
```diff
@@ -692,3 +692,51 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Unfold")>]
         let unfold<'T, 'State> (generator:'State -> ('T*'State) option) (state:'State) = Microsoft.FSharp.Primitives.Basics.List.unfold generator state
+
+        [<CompiledName("RemoveAt")>]
+        let removeAt (index: int) (source: 'T list) : 'T list =
+            if index < 0 || index >= source.Length then invalidArg "index" "index must be within bounds of the array"
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/list.fs:698]: The `.Length` requires an iteration of the list.  I think it's non-optimal

File: src/fsharp/FSharp.Core/list.fs [hunk 9]
```diff
@@ -692,3 +692,51 @@ namespace Microsoft.FSharp.Collections
...
+
+            let before, tail = Microsoft.FSharp.Primitives.Basics.List.splitAt index source
+            let _, after = Microsoft.FSharp.Primitives.Basics.List.splitAt 1 tail
+            before @ after
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/list.fs:702]: There will be a better implementation using an explicit [ListCollector](https://github.com/dotnet/fsharp/blob/f45ef819b5310de521645dbc68ccbfc57f8edda4/src/fsharp/FSharp.Core/seqcore.fsi#L161) something like this
```fsharp
            let mutable i = 0
            let mutable coll = ListCollector()
            let mutable curr = source
            while i < index do
                  match curr with
                  | [] ->invalidArg "index" "index must be within bounds of the list" 
                  | h::t ->
                      coll.Add(h)
                      curr <- t
                  i <- i + 1
            coll.AddManyAndClose(curr.Tail)
```

File: src/fsharp/FSharp.Core/array.fs (hunk 1 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/array.fs]: This one should be fixed for all the functions now

File: src/fsharp/FSharp.Core/map.fs (hunk 4 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fs]: This should be fixed now

File: src/fsharp/FSharp.Core/map.fs (hunk 5 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fs]: This should be fixed now. I added the note of the duplicate values to the xml doc.

File: src/fsharp/FSharp.Core/map.fsi (hunk 6 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fsi]: I added a xml doc and hunted all the seq types in all the places to switch them to this form

File: src/fsharp/FSharp.Core/seq.fs (hunk 7 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/seq.fs]: should be fixed now

**Comment by uxsoft** [comment]
@dsyme I found out that the Map.Values implementation collides with internal Map.Values extension in illib.fs - MapAutoOpens module which returns a list instead of seq. This is the reason the solution right now doesn't build.

Since it's not part of the spec you provided I'll remove it from this PR. That said I see it's something that IS used, so let me know if you'd like me to take a look at it in a separate PR.

File: src/fsharp/FSharp.Core/array.fs [hunk 10]
```diff
@@ -1270,6 +1270,96 @@ namespace Microsoft.FSharp.Collections
...
+            
+            let length = source.Length - 1
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+
+            for i in 0..length do
```
Comment (En3Tho) [inline src/fsharp/FSharp.Core/array.fs:1281]: I don't see why enumeration is needed, especially with conditional inside. Arrays have index based access. You just check the index, create new array and then do 2 Span/Buffer/Array (depends on what's available now in FSharl.Core) Copy calls and that's it. I would like you to change array implementations using this approach as the one you did simply does unnecessary things and thus not optimal.

**Review by En3Tho — CHANGES_REQUESTED** [review]

File: src/fsharp/FSharp.Core/map.fs (hunk 5 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fs]: So I reverted the change because the Map.Values on the type clashed with an internal implementation in illib.fs. I left Map.values in the module with the only implementation.

File: src/fsharp/FSharp.Core/list.fs (hunk 9 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/list.fs:702]: I rewrote all the list implementations to use ListCollector instead

File: src/fsharp/FSharp.Core/array.fs (hunk 3 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/array.fs:1276]: I thought about this and if my code makes an operation that doesn't make sense I'd like to know about it. I checked and that's consistent with other .NET collections.

File: src/fsharp/FSharp.Core/list.fs (hunk 8 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/list.fs]: I rewrote all the list implementations to use ListCollector instead

File: src/fsharp/FSharp.Core/list.fs [hunk 11]
```diff
@@ -692,3 +692,51 @@ namespace Microsoft.FSharp.Collections
...
+            let _, after = Microsoft.FSharp.Primitives.Basics.List.splitAt count tail
+            before @ after
+    
+        [<CompiledName("UpdateAt")>]
+        let updateAt (index: int) (value: 'T) (source: 'T list) : 'T list =
+            if index < 0 || index >= source.Length then invalidArg "index" "index must be within bounds of the array"
+
+            let before, tail = Microsoft.FSharp.Primitives.Basics.List.splitAt index source
+            let _, after = Microsoft.FSharp.Primitives.Basics.List.splitAt 1 tail
+            before @ [ value ] @ after
```
Comment (Happypig375) [inline src/fsharp/FSharp.Core/list.fs]: Why aren't you using `mapi`?

File: src/fsharp/FSharp.Core/array.fs [hunk 12]
```diff
@@ -1308,25 +1303,27 @@ namespace Microsoft.FSharp.Collections
             let length = source.Length
             let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
 
-            for i in 0..length - 1 do
-                result.[i] <-
-                    if i = index then value
-                    else source.[i]
+            Array.Copy(source, result, index)
+            result.[index] <- value
+            Array.Copy(source, index + 1, result, index + 1, length - index - 1)
```
Comment (Happypig375) [inline src/fsharp/FSharp.Core/array.fs:1308]: Can be done with just one copy and then an assign

File: tests/FSharp.Core.UnitTests/SurfaceArea.fs [hunk 13]
```diff
@@ -9,2468 +9,2491 @@ type SurfaceAreaTest() =
     [<Fact>]
     member this.VerifyArea() =
         let expected = @"
-Microsoft.FSharp.Collections.Array2DModule: Int32 Base1[T](T[,])
```
Comment (dsyme) [inline tests/FSharp.Core.UnitTests/SurfaceArea.fs:12]: Do you know why this diff turned out so large?  It makes it hard to review?

File: src/fsharp/FSharp.Core/map.fsi [hunk 14]
```diff
@@ -85,6 +85,9 @@ namespace Microsoft.FSharp.Collections
         /// <returns><c>true</c> if the value is present, <c>false</c> if not.</returns>
         member TryGetValue: key:'Key * [<System.Runtime.InteropServices.Out>] value:byref<'Value> -> bool
 
+        /// <summary>The keys in the map.</summary>
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/map.fsi:88]: Please add property `.Values`

File: src/fsharp/FSharp.Core/seq.fs [hunk 15]
```diff
@@ -1488,3 +1488,88 @@ namespace Microsoft.FSharp.Collections
...
+                    i <- i + 1
+                if i <= index then invalidArg "index" "index must be within bounds of the array"
+            }
+    
+        [<CompiledName("UpdateManyAt")>]
+        let updateManyAt (index: int) (values: seq<'T>) (source: seq<'T>) : seq<'T> =
+            if index < 0 then invalidArg "index" "index must be within bounds of the array"
+            seq {
+                let mutable i = 0
+                let sourceEnumerator = source.GetEnumerator()
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/seq.fs:1534]: This should be `use sourceEnumerator = source.GetEnumerator()`.  However I think we will be removing the ManyAt functions, I'll write a separate note about that.

**Review by dsyme — CHANGES_REQUESTED** [review]
This is looking good. 

Based on the suggestion discussion and looknig at the implementation I think we should at least drop `updateManyAt` across the board.  It is just too corner-case.  

I'll discuss on the suggestion further

File: src/fsharp/FSharp.Core/map.fsi (hunk 14 above)
Comment (dsyme) [reply src/fsharp/FSharp.Core/map.fsi:88]: I saw your comment from earlier:

> @dsyme I found out that the Map.Values implementation collides with internal Map.Values extension in illib.fs - MapAutoOpens module which returns a list instead of seq. This is the reason the solution right now doesn't build.

Yes the one in `illib.fs` should be renamed or removed as part of this PR.

However I have some concerns about Keys and Values, I'll discuss on the suggestion thread

File: src/fsharp/FSharp.Core/seq.fs (hunk 15 above)
Comment (dsyme) [reply src/fsharp/FSharp.Core/seq.fs:1534]: See https://github.com/fsharp/fslang-suggestions/issues/1047#issuecomment-894227703

File: tests/FSharp.Core.UnitTests/SurfaceArea.fs (hunk 13 above)
Comment (uxsoft) [reply tests/FSharp.Core.UnitTests/SurfaceArea.fs:12]: I have no idea. I have replaced the surface output from my PC and how the tests runs fine on both my mac and my windows machine but the tests in the repo still fail. I have .NET 6 previews installed but I don't think the build system uses them?

I also thought about reverting this commit and just manually adding the added methods but i was not sure if the order was important. 

I'd appreciate any help or suggestions on how to resolve this test.

File: src/fsharp/FSharp.Core/map.fsi (hunk 14 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fsi:88]: Ok, I'll remove the illib.fs implementation and fix all the usages as part of this PR. I'll keep it as seq<'Key> for now but I'm happy to update when decided.

File: src/fsharp/FSharp.Core/array.fs (hunk 10 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/array.fs]: I updated the implementation to use array.copy

File: src/fsharp/FSharp.Core/array.fs (hunk 2 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/array.fs]: Fixed now in the new implementation

**Comment by vzarytovskii** [comment]
I'm going to convert it to draft, as it is still WIP, please 'ready to review' it, once ready.

File: src/fsharp/FSharp.Core/array.fs [hunk 16]
```diff
@@ -1270,6 +1270,88 @@ namespace Microsoft.FSharp.Collections
...
+        [<CompiledName("InsertManyAt")>]
+        let insertManyAt (index: int) (values: seq<'T>) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index > source.Length then invalidArg "index" "index must be within bounds of the array"
+            
+            let valuesArray = Seq.toArray values
+            let length = source.Length + valuesArray.Length
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+            
+            Array.Copy(source, result, index)
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs:1350]: Avoid the calls to Array.Copy when the copy length is zero

File: src/fsharp/FSharp.Core/array.fs [hunk 17]
```diff
@@ -1270,6 +1270,88 @@ namespace Microsoft.FSharp.Collections
...
+        
+        [<CompiledName("InsertAt")>]
+        let insertAt (index: int) (value: 'T) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index > source.Length then invalidArg "index" "index must be within bounds of the array"
+            
+            let length = source.Length + 1
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+
+            Array.Copy(source, result, index)
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs:1336]: Avoid the calls to Array.Copy when the copy length is zero

File: src/fsharp/FSharp.Core/map.fs [hunk 18]
```diff
@@ -819,6 +821,70 @@ and
...
+
+        member _.IsReadOnly = true
+
+        member _.Count = parent.Count
+        
+    interface IEnumerable<'Value> with
+        member _.GetEnumerator() =
+            (seq { for item in parent do item.Value}).GetEnumerator()
+            
+    interface Collections.IEnumerable with
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/map.fs:884]: Use `open System.Collections` at the top

File: src/fsharp/FSharp.Core/map.fsi (hunk 14 above)
Comment (dsyme) [reply src/fsharp/FSharp.Core/map.fsi:91]: Great!

**Review by dsyme — CHANGES_REQUESTED** [review]
I think just benchmarks and removing `updateManyAt` is all that's now required.

It may be good to get the F# community to help perf test variations.

File: src/fsharp/FSharp.Core/array.fs [hunk 19]
```diff
@@ -1270,6 +1270,88 @@ namespace Microsoft.FSharp.Collections
...
                 Microsoft.FSharp.Primitives.Basics.Array.subUnchecked 0 count' array
 
+        [<CompiledName("RemoveAt")>]
+        let removeAt (index: int) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index >= source.Length then invalidArg "index" "index must be within bounds of the array"
+            
+            let length = source.Length - 1
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+            Array.Copy(source, result, index)
```
[the remark this answers is not in the document]
Comment (En3Tho) [reply src/fsharp/FSharp.Core/array.fs:1280]: Well, I like to quote Steven Toub here: it's harmful to add such checks because they actually regress perf in most used scenario, which is when array is not empty.

[the remark this answers is not in the document]
Comment (dsyme) [reply src/fsharp/FSharp.Core/array.fs:1280]: Hmmm I don't know if it's the least used case.  I mean, it could be that this is often used on small lists where the insert is at front or back.  And Array.Copy will have a bunch of null checks, overlap checks etc. (e.g. copying same array to itself) I expect. 

Anyway this is what perf benchmarking is for, to check the difference :)  The benchmarks should I guess include both large array and small array tests (size 0, 1, 2, 3), including insert-at-start and insert-at-end

**Comment by dsyme** [comment]
@vzarytovskii We're getting lots of these failures
```
   System.ObjectDisposedException : Cannot write to a closed TextWriter.
```

File: src/fsharp/FSharp.Core/seq.fs (hunk 15 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/seq.fs]: removed

File: src/fsharp/FSharp.Core/array.fs (hunk 19 above)
[the remark this answers is not in the document]
Comment (uxsoft) [reply src/fsharp/FSharp.Core/array.fs]: i added the check for all array.copy calls

File: src/fsharp/FSharp.Core/array.fs [hunk 20]
```diff
@@ -1270,6 +1270,88 @@ namespace Microsoft.FSharp.Collections
...
+        
+        [<CompiledName("RemoveManyAt")>]
+        let removeManyAt (index: int) (count: int) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index > source.Length - count then invalidArg "index" "index must be within bounds of the array"
+            
+            let length = source.Length - count
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+
+            Array.Copy(source, result, index)
```
[the remark this answers is not in the document]
Comment (uxsoft) [reply src/fsharp/FSharp.Core/array.fs]: i added the check for all array.copy calls

File: src/fsharp/FSharp.Core/map.fs [hunk 21]
```diff
@@ -687,6 +687,8 @@ type Map<[<EqualityConditionalOn>]'Key, [<EqualityConditionalOn; ComparisonCondi
     member m.ToArray() =
         MapTree.toArray tree
 
+    member m.Keys = seq { for kvp in m -> kvp.Key }
```
[the remark this answers is not in the document]
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fs]: it's now an ICollection<'Key> I also improved the implementation based on how Dictionary<'T> does it

File: src/fsharp/FSharp.Core/map.fs [hunk 22]
```diff
@@ -933,3 +935,9 @@ module Map =
     [<CompiledName("Count")>]
     let count (table: Map<_, _>) =
         table.Count
+
+    [<CompiledName("Keys")>]
+    let keys (table: Map<_, _>) = table.Keys
+
+    [<CompiledName("Values")>]
+    let values (table: Map<_, _>) = seq { for kvp in table -> kvp.Value }
```
[the remark this answers is not in the document]
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fs]: removed

File: tests/benchmarks/MicroPerf/Collections.fs [hunk 23]
```diff
@@ -0,0 +1,298 @@
...
+        async {
+            List.init x.Length (fun _ -> 0)
+            |> List.removeAt 0 
+            |> ignore
+        }
+        |> Async.StartAsTask
+
+    [<Benchmark>]
+    member x.ListRemoveAtEnd() =
+        async {
```
Comment (dsyme) [inline tests/benchmarks/MicroPerf/Collections.fs:29]: Is the use of tasks to run the benchmarks in parallel?  It's a bit odd as there will be allocation overhad from async and StartAsTask

File: tests/benchmarks/MicroPerf/Collections.fs [hunk 24]
```diff
@@ -0,0 +1,298 @@
...
+            Array.init x.Length (fun _ -> 0)
+            |> Array.removeManyAt (x.Length - 11) 10 
+            |> ignore
+        }
+        |> Async.StartAsTask
+        
+    [<Benchmark>]
+    member x.ArrayInsertAtBeginning() =
+        async {
+            Array.init x.Length (fun _ -> 0)
```
Comment (dsyme) [inline tests/benchmarks/MicroPerf/Collections.fs:148]: It would be best if the benchmarks only measured the operation (Array.insertAt) and not things like Array.init.

File: tests/benchmarks/MicroPerf/Collections.fs [hunk 25]
```diff
@@ -0,0 +1,298 @@
...
+            |> ignore
+        }
+        |> Async.StartAsTask
+        
+    /// Seq
+    [<Benchmark>]
+    member x.SeqRemoveAtBeginning() =
+        async {
+            Seq.init x.Length (fun _ -> 0)
+            |> Seq.removeAt 0
```
Comment (dsyme) [inline tests/benchmarks/MicroPerf/Collections.fs:204]: For Seq I think it's particularly important to only measure the operation - it may be `Seq.init` is doing more allocation that we don't want to measure.

So I think use `Seq.removeAt` etc. on a fixed, pre-allocated list or array

**Review by dsyme — CHANGES_REQUESTED** [review]
This is looking great, just a few comments on the benchmarking

File: tests/benchmarks/MicroPerf/Collections.fs (hunk 23 above)
Comment (uxsoft) [reply tests/benchmarks/MicroPerf/Collections.fs]: Honestly, I just copied what the other benchmarks were doing without giving it much thought. I removed the async/tasks now.

File: tests/benchmarks/MicroPerf/Collections.fs (hunk 24 above)
Comment (uxsoft) [reply tests/benchmarks/MicroPerf/Collections.fs]: I changed all the methods to use an array/list which is pre-initialised.

File: tests/benchmarks/MicroPerf/Collections.fs (hunk 25 above)
Comment (uxsoft) [reply tests/benchmarks/MicroPerf/Collections.fs]: I also changed the Seq.toList to Seq.iter ignore, hopefully, that'll be more accurate.

File: tests/benchmarks/MicroPerf/Collections.fs (hunk 23 above)
Comment (dsyme) [reply tests/benchmarks/MicroPerf/Collections.fs]: Yup I hadn't reviewed those - thanks!

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
I've added an RFC for this here: https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1113-insert-remove-update-functions.md

File: src/fsharp/FSharp.Core/map.fsi [hunk 26]
```diff
@@ -83,7 +83,16 @@ namespace Microsoft.FSharp.Collections
         /// <param name="value">A reference to the output value.</param>
         ///
         /// <returns><c>true</c> if the value is present, <c>false</c> if not.</returns>
-        member TryGetValue: key:'Key * [<System.Runtime.InteropServices.Out>] value:byref<'Value> -> bool
+        member TryGetValue: key: 'Key * [<System.Runtime.InteropServices.Out>] value: byref<'Value> -> bool
+        
+        /// <summary>The empty map.</summary>
+        static member Empty: Map<'Key,'Value> when 'Key: comparison
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/map.fsi:89]: I hadn't noticed this addition - `Empty` should be removed, as the function `Map.empty` is preferred, and we don't want both showing in autocomplete lists

**Comment by dsyme** [comment]
@uxsoft I updated to 

1. Add Map.values
2. Remove Map.Empty (in favour of existing Map.empty)

File: src/fsharp/FSharp.Core/map.fsi (hunk 26 above)
Comment (uxsoft) [reply src/fsharp/FSharp.Core/map.fsi]: I remember now, I moved this from the additions in illib.fs. I'll refactor the illib.fs code to refer to Map.empty so the build is green again

Comment (dsyme) [reply src/fsharp/FSharp.Core/map.fsi]: Thanks!

**Comment by dsyme** [comment]
@uxsoft Thank you for this contribution!

**Comment by uxsoft** [comment]
Thank you for your guidance!

**Comment by OkkeHendriks** [comment]
I am not sure if this is the place to ask this, as this is already merged.
However, I see that at the `toSeq` [function](https://github.com/uxsoft/fsharp/blob/2be0e670c4bcdaa69f102da03380fe3fd766a371/src/fsharp/FSharp.Core/map.fsi#L154) the summary states that `The sequence will be ordered by the keys of the map.`. This is also the case for `Map.keys` and `Map.values`, so I think it is an interesting property to note within the summary of those.

Also the tests should verify this if this is added.

**Comment by dsyme** [comment]
> This is also the case for Map.keys and Map.values, so I think it is an interesting property to note within the summary of those.

Yes, I agree.  It's also interesting to think whether things like "get me the middle key in the ordering of keys" will run efficiently or not

    let midKey =
        let keys = map.Keys
        keys.[keys.Length/2]

I haven't checked. Probably not

**Comment by uxsoft** [comment]
I can open a new PR with these small changes if you'd like

**Comment by OkkeHendriks** [comment]
> It's also interesting to think whether things like "get me the middle key in the ordering of keys" will run efficiently or not.

Correct me if I'm wrong but I think that is not even possible, directly, as `ICollection` does not implement `Item`:
image

So to get the middle one would have to `MoveNext()` through the sequence.

**Comment by dsyme** [comment]
> Correct me if I'm wrong but I think that is not even possible, directly, as ICollection does not implement Item:

Yes you're correct, thanks

**Comment by dsyme** [comment]
> I can open a new PR with these small changes if you'd like

Yes, please make the additions to the doc comments

File: src/fsharp/NameResolution.fs [hunk 27]
```diff
@@ -4320,11 +4320,12 @@ let rec ResolvePartialLongIdentPrim (ncenv: NameResolver) (nenv: NameResolutionE
...
            | OpenQualified ->
                nenv.eUnqualifiedItems.Values
-               |> List.filter (function
+               |> Seq.filter (function
                    | Item.UnqualifiedType _ -> false
                    | Item.Value v -> not v.IsMember
                    | _ -> true)
-               |> List.filter (ItemIsUnseen ad g ncenv.amap m >> not)
+               |> Seq.filter (ItemIsUnseen ad g ncenv.amap m >> not)
+               |> Seq.toList
```
Comment (auduchinok) [inline src/fsharp/NameResolution.fs:4328]: Why this change was needed?

File: src/fsharp/absil/illib.fs [hunk 28]
```diff
@@ -1127,11 +1127,9 @@ type LayeredMap<'Key, 'Value when 'Key : comparison> = Map<'Key, 'Value>
 [<AutoOpen>]
 module MapAutoOpens =
     type Map<'Key, 'Value when 'Key : comparison> with
-
+        
         static member Empty : Map<'Key, 'Value> = Map.empty
-
-        member x.Values = [ for KeyValue(_, v) in x -> v ]
```
Comment (auduchinok) [inline src/fsharp/absil/illib.fs:1133]: @dsyme Do we expect all FCS clients to use FSharp.Core preview now? I don't think it's a good idea to tie FCS and FSharp.Core previews like this.

The tooling now shows errors (https://github.com/dotnet/fsharp/issues/12142) and it seems we won't be able to run FCS with a stable FSharp.Core version anymore?

Comment (dsyme) [reply src/fsharp/absil/illib.fs:1133]: @auduchinok Hmmmm could you please open a separate issue about this.  I agree public non-pre-release FCS should indeed only use a public non-pre-release FCS.  

However if FCS in pre-release and the release plan is matching FSharp.Core than it is ok

File: src/fsharp/NameResolution.fs (hunk 27 above)
Comment (dsyme) [reply src/fsharp/NameResolution.fs:4328]: I doubt it was necessary

Comment (uxsoft) [reply src/fsharp/NameResolution.fs:4328]: There was an internal implementation of keys/values which returned a list and the implementation was counting with that.

Since the new public implementation of keys/values is an ICollection I had to change this to take seq instead of List otherwise it would not compile.
